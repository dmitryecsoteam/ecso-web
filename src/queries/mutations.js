import gql from 'graphql-tag';

export const SIGNUP_USER = gql`
mutation signupUser($email: String, $password: String, $name: String){
    signupUser (email: $email, password: $password, name: $name) {
        token
    }
}
`;

export const SIGNIN_USER = gql`
mutation signinUser($email: String, $password: String){
    signinUser (email: $email, password: $password) {
        token
    }
}
`;

export const DELETE_NOTIFICATION = gql`
mutation deleteNotification($id: String){
    deleteNotification (id: $id) {
        notifications {
            origin { nameEn }
            destination { nameEn }
            travelId
            date
            priceHotelLast
            priceAirplaneLast
        }
    }
}
`;

export const ADD_NOTIFICATION = gql`
mutation addNotification($id: String, $date: String, $origin: Int, $destination: Int) {
    addNotification (id: $id, date: $date, origin: $origin, destination: $destination) {
        notifications {
            travelId
            date 
            priceAirplaneLast
            priceHotelLast
        }
    }
}
`;