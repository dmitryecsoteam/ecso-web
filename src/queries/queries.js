import gql from 'graphql-tag';

export const ORIGIN_INPUT_SEARCH = gql`
query originStartsWith($name: String){
    originStartsWith (name: $name) {
        name,
        country
    }
}
`