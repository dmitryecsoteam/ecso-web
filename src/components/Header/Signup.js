import React from 'react';
import { Mutation } from 'react-apollo';
import SHA256 from 'crypto-js/sha256';

import InputWithErrorTooltip from '../MainPage/InputWithErrorTooltip';
import { SIGNUP_USER } from '../../queries/mutations';


export default class Signup extends React.Component {
  state = {
    email: '',
    name: '',
    password: '',
    passwordConfirmation: '',
    emailError: false,
    emailErrorText: '',
    nameError: false,
    nameErrorText: '',
    passwordError: false,
    passwordErrorText: '',
    passwordConfirmationError: false,
    passwordConfirmationErrorText: '',
  }

  onInputChange = (event) => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    });
  }

  onFormSubmit = async (event, signupUser) => {
    event.preventDefault();

    const { email, name, password, passwordConfirmation } = this.state;

    let emailError = false,
      emailErrorText = '',
      nameError = false,
      nameErrorText = '',
      passwordError = false,
      passwordErrorText = '',
      passwordConfirmationError = false,
      passwordConfirmationErrorText = '';

    if (!email) {
      emailError = true;
      emailErrorText = 'Enter your email';
    }

    if (!name) {
      nameError = true;
      nameErrorText = 'Enter your name';
    }

    if (!password) {
      passwordError = true;
      passwordErrorText = 'Enter password';
    }

    if (!passwordConfirmation) {
      passwordConfirmationError = true;
      passwordConfirmationErrorText = 'Confirm your password'
    } else {
      if (password !== passwordConfirmation) {
        passwordConfirmationError = true;
        passwordConfirmationErrorText = 'Passwords didn\'t match';
      }
    }

    this.setState({
      emailError,
      emailErrorText,
      nameError,
      nameErrorText,
      passwordError,
      passwordErrorText,
      passwordConfirmationError,
      passwordConfirmationErrorText,
    });



    if (!(emailError || nameError || passwordError || passwordConfirmationError)) {
      try {

        const { data } = await signupUser();
        console.log(data);

        localStorage.setItem('token', data.signupUser.token);

        this.props.fetchUser();

        if (this.props.closeModal) this.props.closeModal();

      } catch (e) {

        console.log(e)
        this.setState({
          emailError: true,
          emailErrorText: e.message.replace('GraphQL error: ', '')
        });

      }
    }


  }

  render() {

    const {
      email,
      name,
      password,
      passwordConfirmation,
      emailError,
      emailErrorText,
      nameError,
      nameErrorText,
      passwordError,
      passwordErrorText,
      passwordConfirmationError,
      passwordConfirmationErrorText,
    } = this.state;

    const passwordSHA = SHA256(password).toString();

    return (
      <div>
        <h2 className="signup__title">Signup with email</h2>

        <Mutation mutation={SIGNUP_USER} variables={{ email, password: passwordSHA, name }}>

          {(signupUser, { data, loading, error }) => {

            return (
              <form
                onSubmit={e => this.onFormSubmit(e, signupUser)}
                autoComplete="off"
              >

                <InputWithErrorTooltip
                  error={emailError}
                  errorText={emailErrorText}
                >
                  <input
                    type="text"
                    name="email"
                    value={email}
                    placeholder="Email"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={nameError}
                  errorText={nameErrorText}
                >
                  <input
                    type="text"
                    name="name"
                    value={name}
                    placeholder="Name"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={passwordError}
                  errorText={passwordErrorText}
                >
                  <input
                    type="password"
                    name="password"
                    value={password}
                    placeholder="Password"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <InputWithErrorTooltip
                  error={passwordConfirmationError}
                  errorText={passwordConfirmationErrorText}
                >
                  <input
                    type="password"
                    name="passwordConfirmation"
                    value={passwordConfirmation}
                    placeholder="Confirm password"
                    className="signup__input"
                    onChange={this.onInputChange}
                  />
                </InputWithErrorTooltip>

                <button className="search-form__submit-btn">Submit</button>
              </form>);
          }}

        </Mutation>

      </div>
    );
  }
}