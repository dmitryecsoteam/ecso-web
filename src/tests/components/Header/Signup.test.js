import React from 'react';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import { GraphQLError } from 'graphql';

import Signup from '../../../components/Header/Signup';
import { SIGNUP_USER } from '../../../queries/mutations';


const waitForAsync = () => new Promise(resolve => setTimeout(resolve, 0));

let wrapper;
beforeEach(() => {
    const mocks = [{
        request: {
            query: SIGNUP_USER,
            variables: { email: 'test@mail.com', password: 'asdas', name: 'Alex' }
        },
        result: { data: { signupUser: { token: 'qwerty12' } } }
    }, {
        request: {
            query: SIGNUP_USER,
            variables: { email: 'exists@mail.com', password: 'asdas', name: 'Alex' }
        },
        result: {
            errors: [new GraphQLError('User with email exists@mail.com already exists')],
          }
    }];

    wrapper = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Signup />
        </MockedProvider>
    );
});




test('should render Signup component', () => {
    expect(wrapper.find(Signup)).toMatchSnapshot();
});

test('should signup new user and return token', async () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'asdas' } });

    wrapper.find('form').simulate('submit');
    await waitForAsync();

    // wrapper.update();
    // console.log(wrapper.find(Signup).state())
});

test('should not signup user with existing email', async () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'exists@mail.com' } });
    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'asdas' } });

    wrapper.find('form').simulate('submit');
    await waitForAsync();


    const signup = wrapper.find(Signup);
    //Wait for setState to update
    await waitForAsync();
    expect(signup.state('emailError')).toBeTruthy();
    expect(signup.state('emailErrorText')).toEqual('User with email exists@mail.com already exists');
});

test('should not submit form without email', () => {

    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'asdas' } });

    wrapper.find('form').simulate('submit');

    const signup = wrapper.find(Signup);
    expect(signup.state('emailError')).toBeTruthy();
    expect(signup.state('emailErrorText')).toEqual('Enter your email');
});

test('should not submit form without name', () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'asdas' } });

    wrapper.find('form').simulate('submit');

    const signup = wrapper.find(Signup);
    expect(signup.state('nameError')).toBeTruthy();
    expect(signup.state('nameErrorText')).toEqual('Enter your name');
});

test('should not submit form without password', () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'asdas' } });

    wrapper.find('form').simulate('submit');

    const signup = wrapper.find(Signup);
    expect(signup.state('passwordError')).toBeTruthy();
    expect(signup.state('passwordErrorText')).toEqual('Enter password');
});

test('should not submit form without passwordConfirmation', () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });

    wrapper.find('form').simulate('submit');

    const signup = wrapper.find(Signup);
    expect(signup.state('passwordConfirmationError')).toBeTruthy();
    expect(signup.state('passwordConfirmationErrorText')).toEqual('Confirm your password');
});

test('should not submit form if passwords didn\'t match', () => {

    wrapper.find('[name="email"]').simulate('change', { target: { name: 'email', value: 'test@mail.com' } });
    wrapper.find('[name="name"]').simulate('change', { target: { name: 'name', value: 'Alex' } });
    wrapper.find('[name="password"]').simulate('change', { target: { name: 'password', value: 'asdas' } });
    wrapper.find('[name="passwordConfirmation"]').simulate('change', { target: { name: 'passwordConfirmation', value: 'qqqqq' } });

    wrapper.find('form').simulate('submit');

    const signup = wrapper.find(Signup);
    expect(signup.state('passwordConfirmationError')).toBeTruthy();
    expect(signup.state('passwordConfirmationErrorText')).toEqual('Passwords didn\'t match');
});