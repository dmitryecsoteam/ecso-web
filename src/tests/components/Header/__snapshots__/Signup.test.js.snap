// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Signup component 1`] = `
<Signup>
  <div>
    <h2
      className="signup__title"
    >
      Signup with email
    </h2>
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": Object {
                "kind": "Name",
                "value": "signupUser",
              },
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "email",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "email",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "password",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "password",
                          },
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "value": Object {
                          "kind": "Variable",
                          "name": Object {
                            "kind": "Name",
                            "value": "name",
                          },
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "signupUser",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "token",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "email",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "password",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 161,
            "start": 0,
          },
        }
      }
      variables={
        Object {
          "email": "",
          "name": "",
          "password": "",
        }
      }
    >
      <form
        autoComplete="off"
        onSubmit={[Function]}
      >
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <_default
          error={false}
          errorText=""
        >
          <div
            className="search-form__input-container"
          >
            <div
              className="search-form__label-input"
            >
              <input
                className="signup__input"
                name="passwordConfirmation"
                onChange={[Function]}
                placeholder="Confirm password"
                type="password"
                value=""
              />
            </div>
            <div
              className="search-form__error"
            >
              <span
                className="search-form__error-text"
              />
            </div>
          </div>
        </_default>
        <button
          className="search-form__submit-btn"
        >
          Submit
        </button>
      </form>
    </Mutation>
  </div>
</Signup>
`;
